File: system-test/dotnet/Dsl.Core/UseCase/UseCases/BrowseCoupons.cs
using Driver.Port;
using Driver.Port.Dtos;
using Driver.Port.Dtos.Error;
using Dsl.Core.Shared;
using Dsl.Core.UseCase.UseCases.Base;

namespace Dsl.Core.UseCase.UseCases;

public class BrowseCoupons : BaseMyShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification>
{
    public BrowseCoupons(IMyShopDriver driver, UseCaseContext context)
        : base(driver, context)
    {
    }

    public override async Task<MyShopUseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification>> Execute()
    {
        var result = await _driver.BrowseCouponsAsync();

        return new MyShopUseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification>(
            result,
            _context,
            (response, ctx) => new BrowseCouponsVerification(response, ctx));
    }
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/support/core/usecase/usecases/BrowseCoupons.java
package com.mycompany.myshop.backend.support.core.usecase.usecases;

import com.mycompany.myshop.backend.core.dtos.BrowseCouponsResponse;
import com.mycompany.myshop.backend.support.BackendDriver;
import com.mycompany.myshop.backend.support.core.shared.UseCaseResult;
import com.mycompany.myshop.backend.support.core.usecase.usecases.base.BaseMyShopUseCase;
import org.springframework.http.HttpStatus;

/** {@code GET /api/coupons} — always {@code 200 OK}; there is no rejection path to state. */
public class BrowseCoupons extends BaseMyShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification> {

    public BrowseCoupons(BackendDriver driver) {
        super(driver);
    }

    @Override
    public UseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification> execute() {
        var response = driver.browseCoupons();

        return new UseCaseResult<>(
            response.getStatusCode(),
            HttpStatus.OK,
            null,
            response::getBody,
            null,
            BrowseCouponsVerification::new);
    }
}
